### ISSUE TITLE
URL search should only ever show a valid entity or empty input when unfocused

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#4427]
Per https://www.figma.com/file/VILRieNSLg2DOVyEgFBkXe/%5BMVP%5D-Unified-Dashboard?node-id=1330%3A9080, a few late updates have been made to give the URL search and entity dashboard header for the `unifiedDashboard` a better user experience. This is the second out of three issues which replaces #4147. The others are:
* #4426
* #4428

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

With the `unifiedDashboard` feature flag enabled:
* When unfocused, it should only ever be possible for the URL search to show either an empty state or a valid search result. Specifically:
    * When entering something without actually choosing a result from the autocomplete, the URL search should empty itself again upon unfocusing (e.g. clicking away, tabbing out, etc.).
    * When clicking into the URL search when there's already a selection in there (only possible via the change in #4426), if the user doesn't type anything and then unfocuses, that selection should remain. (If the user starts to type anything, effectively erasing that selection, and then unfocuses, the URL search should empty itself as stated above.)
    * The autocomplete dropdown should only ever be shown if the URL search has focus. In other words, when unfocusing, it should no longer be shown.

## Implementation Brief

* In `assets/js/components/PostSearcherAutoSuggest.js`:
  * Add a new property `setIsActive`.
    * Add a new callback function `onFocus` that calls `setIsActive( true )`.
    * Add a new callback function `onBlur` that calls `setIsActive( false )` and resets the `searchTerm` value back to the current entity title if it is not empty otherwise to an empty string when nothing is selected (#4465 must be merged before this can be implemented).
      * This might require doing it in a `setTimeout` because `onblur` event will be triggered sooner than `onselect` callback.
    * Add new properties `onFocus={ onFocus }` and `onBlur={ onBlur }` to the `ComboboxInput` element.
  * Add a new property `hideDropdown`.
    * Update the autocomplete dropdown not to render if the `hideDropdown` property is not truthy. 
* In `assets/js/components/EntitySearchInput.js`
  * Add a new state variable `active`.
  * Update the `PostSearcherAutoSuggest` element to pass the setter function for the `active` state as the `setIsActive` property and `! active` value as the `hideDropdown` property.
  * Update the `ProgressBar` element not to render if the `active` value is falsy.

### Test Coverage

* N/A

## QA Brief

* With `unifiedDashboard` flag enabled:
    * On Main Dashboard
        * Type some terms without selecting anything.
        * Unfocus by clicking outside the search bar and using `ESC` key. 
        * The search term should get cleared. 
        * The suggestion box should disappear. 
    * On Entity Dashboard.
       * Type new search tern replacing the current entity title that is auto populated. 
       * Upon unfocus, the new term should get replaced with current entity title and suggestions should disappear. 

## Changelog entry

* Update entity search field to only show a valid result when unfocused.

**Comment by eugene-manuilov** [comment]
> If the user starts to type anything, effectively erasing that selection, and then unfocuses, the URL search should empty itself as stated above.

@felixarntz I have added IB but it doesn't include implementation details for this ^ part of AC. I am wondering whether we need to implement this because it can easily become very annoying if we keep erasing the entered search query when the user clicks away (by accident or not, doesn't matter). I think that we need to reset it back to the current entity title, only if the user has cleared the search input and gone away. What do you think?

**Comment by felixarntz** [comment]
@eugene-manuilov Emptying the search on unfocusing was defined by @samitron7, so I assume this is a reasonable UX pattern here. I also agree with that, since it's an autocomplete, which is almost like a dropdown. So if you enter a random text and don't select any of the options, the input is invalid, so it should be emptied if you move away from the field.

You can only enter something to then choose one of the predefined values, like in a `select`. You can't just put "anything" - which is different in regular search fields, but what we have here isn't a classical search field.

**Comment by eugene-manuilov** [comment]
@felixarntz IB is updated. Does it look good to you?

**Comment by felixarntz** [comment]
IB ✅

**Comment by kuasha420** [comment]
@aaemnnosttv addressed the issue on the Code Review. Please have a look. Also ignore the amazing milestone, it was added by one of the kittens while I went to grab water. 

![kitten-on-keyboard](https://user-images.githubusercontent.com/6187401/146034708-9a35b98d-8482-4153-8e66-a1c4301027a4.jpg)

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@kuasha420 two observations Id like to highlight.

1. When you unfocus (i.e. pressing Esc key) the URL Search button is highlighted as if its active.  Is this expected?  

Here's a screenshot to highlight it. 

https://user-images.githubusercontent.com/73545194/146794768-56c82c8d-62af-457a-b18d-0bea012f3d83.mp4

2. On the Entity Dashboard, I see the current entity title auto populated. When I click in the input field to type a new search term, should the title be automatically cleared? 

In the AC it says:

> If the user starts to type anything, effectively erasing that selection..

I am finding that I have to delete the title allowing me then to type in a new search term which is not good UX in my opinion. When I type anything the title remains there.

**Comment by aaemnnosttv** [comment]
> 1. When you unfocus (i.e. pressing Esc key) the URL Search button is highlighted as if its active.  Is this expected?

Yes, it's important for accessibility. See https://github.com/google/site-kit-wp/issues/4493

Regarding the second point, I find the ACs a bit confusing there so we probably need @felixarntz to clarify. I think it means that if the user opens the search field and changes the input without making a new selection, then it should be emptied when unfocusing the input. As it is now it returns back to the current title.

Another thing I want to highlight which looks a bit off is that the search input is still opened when navigating to the entity dashboard from a search result. @felixarntz is this intended? 
![image](https://user-images.githubusercontent.com/1621608/146890283-360fc194-0eab-43db-a333-30051bdc177d.png)

**Comment by felixarntz** [comment]
@wpdarren @aaemnnosttv 
> Another thing I want to highlight which looks a bit off is that the search input is still opened when navigating to the entity dashboard from a search result.

This is intended, per the requirements from #4426. The reason for this is as follows: _If_ we already had a single page application where the entity dashboard would not be a new page load, this is exactly how it should work from a UX perspective. That was one of the requirements that @samitron7 made, and I agree on that. I also agree that it feels a little weird based on that right now there is a full page reload, but when you think about it, in a more integrated SPA that would be the expected behavior. When you type a search query e.g. on Google etc, it doesn't empty itself once you land on the result either.

> I am finding that I have to delete the title allowing me then to type in a new search term which is not good UX in my opinion. When I type anything the title remains there.

I think that is also expected per what I mention above. Again, I agree it may seem a bit off _now_ that we have a full page reload in between, but otherwise that would be the expected behavior, and we should work towards that.

**Comment by eugene-manuilov** [comment]
I also found one regression added by this ticket. I can't select a result with the mouse in the autocomplete list anymore:

https://user-images.githubusercontent.com/873049/146986078-d48888b4-2979-475a-8f5f-bfc0d664dce9.mp4

If I comment out the timeout in the `onBlur` callback, then it works again: https://github.com/google/site-kit-wp/blob/f864bbc681ec362492621a78841b1aea4ba87854/assets/js/components/PostSearcherAutoSuggest.js#L97-L105

**Comment by kuasha420** [comment]
@eugene-manuilov which browser are you using? I have the completely opposite effect. ie. Can not select with mouse when the timeout is removed. (Chrome, linux).

**Comment by eugene-manuilov** [comment]
@kuasha420, I am also on Linux and using Chrome.

**Comment by wpdarren** [comment]
@kuasha420 I have come across two issues that seem to be new since I worked on the QA. 

1. When I start typing in a search term, in console an error appears.  `react_devtools_backend.js:4045 Google Site Kit API Error method:GET datapoint:post-search type:core identifier:search error:"The response is not a valid JSON response."` Initially, `no results found` appears, the error appears in console then a few seconds later the results appear. 

https://user-images.githubusercontent.com/73545194/148047472-740cf85b-12b7-4153-9548-1074abb81356.mp4

2. When I click on a search result in the dropdown, nothing happens. @eugene-manuilov mentioned this [here](https://github.com/google/site-kit-wp/issues/4427#issuecomment-999436123)

**Comment by kuasha420** [comment]
@wpdarren I could not produce the No.1 issue that you've mentioned. Can you test on latest `develop` again? 

As for No.2 issue. When I remove the timeout, it NEVER works for me with pointer. But when the timeout is there, it works for me when I insta-click on the item (No press and hold or long press). But if I press and hold for just a few ms, it doesn't work. 

In short, what happening here is: 

### With timeout:
* Click on a search result 
* `onBlur` fires.
* `isActive` is set to false after 100ms.
* if click is finished (onMouseUp?) within the 100ms, the `onSelectCallback` fires and everything is good.
* if click is not finished within 100ms, the `isActive` gets set to `false`, thus removing the result from the dom tree. Afterwards, click is finished but there is no button in the dom so `onSelectCallback` is not fired, causing the bug.

### Without timeout (in my case):
* Click on a search result 
* `onBlur` fires.
* `isActive` is set to false immediately.
* `onSelectCallback` never fires because the result is removed from the tree and it never gets the click.

I can consistently reproduce this in my local env and on instawp site on latest `develop`. I'm not sure how to best fix this behavior. Any idea? CC @eugene-manuilov @aaemnnosttv 

Cheers.

**Comment by wpdarren** [comment]
@kuasha420 

> @wpdarren I could not produce the No.1 issue that you've mentioned. Can you test on latest develop again?

Sure, I will add that this is inconsistent but the issue still occurs.  

As you can see from the screencast, in the network tab, I started typed the letter `w` and that was status 200, and as I typed `o` the status changed to canceled, but then changes to status 200.  In the console the same error message appears.

`react_devtools_backend.js:4045 Google Site Kit API Error method:GET datapoint:post-search type:core identifier:search error:"You are probably offline."`

I've been able to recreate this on two websites, but it does not occur every time. 

https://user-images.githubusercontent.com/73545194/148377068-e03c35bf-e1fa-472b-9b31-aa30e5440c7b.mp4

**Comment by tofumatt** [comment]
I see the issue and it might be that we need some kind of `onClickOutside` handler or something. I'm going to investigate this and see if there's anything in the ComboBox code that will allow it or if we need to change things around to get the events to fire in the order we want.

The setTimeout solution won't work as it causes "click-and-hold" events not to function, plus causes a reasonable delay.

**Comment by kuasha420** [comment]
@tofumatt Thanks a bunch for looking into it! <3

**Comment by techanvil** [comment]
I've taken a look at this in @tofumatt's absence. Looking at the `ComboBox` code, it doesn't provide any obvious way to implement an `onClickOutside` handler.

I'd suggest the following approach, with a nod to `ComboBox`'s own [`handleBlur`](https://github.com/reach/reach-ui/blob/07d815040f95d192225bac7fff2432fb26f7f90b/packages/combobox/src/index.tsx#L1186) function which makes use of `event.relatedTarget`.

* Add the class `autocomplete__option--result` to the result `ComboboxOption` elements, to distinguish them from the "no results" variant.
* In `onBlur`, only execute the body of the function if `event.relatedTarget` does not contain `autocomplete__option--result`, i.e. only handle the blur if we have **not** clicked on a result.
* Remove the timeout as it's no longer needed.

That is to say:

```diff
diff --git a/assets/js/components/PostSearcherAutoSuggest.js b/assets/js/components/PostSearcherAutoSuggest.js
index bf4e727fb..58f6c44b0 100644
--- a/assets/js/components/PostSearcherAutoSuggest.js
+++ b/assets/js/components/PostSearcherAutoSuggest.js
@@ -94,14 +94,11 @@ export default function PostSearcherAutoSuggest( {
                setIsActive( true );
        }, [ setIsActive ] );
 
-       const onBlur = useCallback( () => {
-               // When a result is selected using pointer, the onBlur event gets fired before the onSelect,
-               // thus removing the dropdown from the tree and missing the click completely. The timeout
-               // ensures that the click event will be fired before isActive is set to false.
-               setTimeout( () => {
+       const onBlur = useCallback( ( event ) => {
+               if ( ! event.relatedTarget?.classList.contains( 'autocomplete__option--result' ) ) {
                        setIsActive( false );
                        setSearchTerm( postTitle.current ?? currentEntityTitle ?? '' );
-               }, 100 );
+               }
        }, [ currentEntityTitle, setIsActive ] );
 
        const onSelectCallback = useCallback(
@@ -264,7 +261,7 @@ export default function PostSearcherAutoSuggest( {
                                                                <ComboboxOption
                                                                        key={ ID }
                                                                        value={ title }
-                                                                       className="autocomplete__option"
+                                                                       className="autocomplete__option autocomplete__option--result"
                                                                />
                                                        ) ) }
                                                </ComboboxList>
```


While perhaps not the most elegant approach, we do examine DOM classes to determine UI state elsewhere in the app, so I think it's probably OK.

**Comment by aaemnnosttv** [comment]
Thanks @techanvil ! This looks reasonable to me; nice work! Would you please open a PR to and from `main` with this change?

**Comment by techanvil** [comment]
Thanks @aaemnnosttv, will do!

**Comment by techanvil** [comment]
PR created :point_up: :+1:

**Comment by techanvil** [comment]
Hi @wpdarren - I've addressed the 2nd point you raised, where clicking on a search result wasn't doing anything.

Regarding your first point relating to the console error and "no results" message, I would suggest raising a separate issue for this, as it's not related to the unfocus behaviour addressed in this ticket, and it doesn't look like something that would have been introduced by the changes in this ticket.

**Comment by tofumatt** [comment]
Oh, yes, agreed on the second point. That seems like an unrelated issue. 👍🏻

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@techanvil I have one observation. 

> On Entity Dashboard:  when typing a term, replacing the current entity title that is auto populated.
Upon unfocus, the new term should get replaced with current entity title and suggestions should disappear.

When I put my cursor into the URL search input field, the current entity title is not deleted, and to replace it, I have to delete it before typing in a new term.  It's not good UX in my opinion. Should this be a new ticket too? 

**Verified:**

_With unifiedDashboard flag enabled:_ 

On Main Dashboard: when typing a term, and without selecting anything, I unfocus by clicking outside the search bar and/or using ESC key, the term is cleared from the search input field. 

On Entity Dashboard:  when typing a term, replacing the current entity title that is auto populated.
Upon unfocus, the new term gets replaced with current entity title and suggestions disappear.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

_With unifiedDashboard flag enabled:_ 

On Main Dashboard: when typing a term, and without selecting anything, I unfocus by clicking outside the search bar and/or using ESC key, the term is cleared from the search input field. 

On Entity Dashboard:  when typing a term, replacing the current entity title that is auto populated.
Upon unfocus, the new term gets replaced with current entity title and suggestions disappear.

https://user-images.githubusercontent.com/73545194/149168745-7db3be2e-faaf-4767-ae6f-e53346a63277.mov

**Comment by aaemnnosttv** [comment]
> When I put my cursor into the URL search input field, the current entity title is not deleted, and to replace it, I have to delete it before typing in a new term. It's not good UX in my opinion. Should this be a new ticket too?

@wpdarren I'm not sure the input should be cleared by focusing it; Google search doesn't do it 😄 

It sounds like you're assuming that the user would always want to replace the entire search input with entirely new terms, but it could be that you want to add/remove/change one word, or just to see the different choices to choose from for the same input which would all be quite frustrating to do if the input was proactively cleared on focus.
